File: src/my_project/medium_problems/from51to100/zigzag_conversion_dp.py
from typing import List 

class Solution:
    def convert(self, s: str, numRows: int) -> str:
    
           
        if numRows == 1 or numRows >= len(s):
            return s

        L = [''] * numRows
        index, step = 0, 1
        
        print(L)

        for x in s:
            L[index] += x
            if index == 0:
                step = 1
            elif index == numRows -1:
                step = -1
            index += step
            print(L)

        return ''.join(L)
    
    
solution = Solution()

print(solution.convert( s = "PAYPALISHIRING", numRows = 3))

File: my_project_typescript/medium_problems_typescript/from1to50/zigZagConversionDP.ts
function convert(s: string, numRows: number): string {

    let lenS: number = s.length

    if (numRows === 1 || numRows >= lenS){
        return s
    }

    let arrayAnswer: Array<string> = Array(lenS).fill('')
    let index: number = 0
    let step: number = 1

    for (let letter of s){
        arrayAnswer[index] += letter 
        if (index === 0){
            step = 1
        }
        else if (index === numRows -1){
            step = -1
        }
        index += step 
    }

    return arrayAnswer.join('')
};